Add Bun streaming server renderer

colinhacks opened the pull request:
<!--
  Thanks for submitting a pull request!
  We appreciate you spending the time to work on these changes. Please provide enough information so that others can review your pull request. The three fields below are mandatory.

  Before submitting a pull request, please make sure the following is done:

  1. Fork [the repository](https://github.com/facebook/react) and create your branch from `main`.
  2. Run `yarn` in the repository root.
  3. If you've fixed a bug or added code that should be tested, add tests!
  4. Ensure the test suite passes (`yarn test`). Tip: `yarn test --watch TestName` is helpful in development.
  5. Run `yarn test --prod` to test in the production environment. It supports the same options as `yarn test`.
  6. If you need a debugger, run `yarn debug-test --watch TestName`, open `chrome://inspect`, and press "Inspect".
  7. Format your code with [prettier](https://github.com/prettier/prettier) (`yarn prettier`).
  8. Make sure your code lints (`yarn lint`). Tip: `yarn linc` to only check changed files.
  9. Run the [Flow](https://flowtype.org/) type checks (`yarn flow`).
  10. If you haven't already, complete the CLA.

  Learn more about contributing: https://reactjs.org/docs/how-to-contribute.html
-->

## Summary

This PR adds a streaming React renderer that's optimized for Bun, and a conditional export in `"react-dom"` that is respected by the `bun` runtime. 

## How did you test this change?

The built versions of `build/node_modules/react-dom/esm/react-dom-server.bun.{development|production}.js` have been tested against Bun's React test suite: https://github.com/oven-sh/bun/blob/main/test/bun.js/react-dom.test.tsx. For the moment these files were manually copied over to the Bun repo, if/when this PR is merged we'll switch to an NPM dependency in our test suite. 

```bash
$ bun wiptest react

react-dom.test.tsx:
✓ React > React.createContext works
✓ server.bun(<a>b</a>) > Response.text()
✓ server.bun(<a>b</a>) > Response.arrayBuffer()
✓ server.bun(<a>b</a>) > Response.blob()
✓ server.bun(<a>b</a>) > readableStreamToText(stream)
✓ server.bun(<a>b</a>) > readableStreamToBlob(stream)
✓ server.bun(<a>b</a>) > readableStreamToArray(stream)
✓ server.bun(<a>b</a>) > readableStreamToArrayBuffer(stream)
✓ server.bun(<a>b</a>) > for await (chunk of stream)
✓ server.bun(<a>b</a>) > for await (chunk of stream) (arrayBuffer)
... more tests
✓ server.browser(<span>Hello World!</span>😋L😋l😋L😋) > Response.arrayBuffer()
✓ server.browser(<span>Hello World!</span>😋L😋l😋L😋) > Response.blob()
✓ server.browser(<span>Hello World!</span>😋L😋l😋L😋) > readableStreamToText(stream)
✓ server.browser(<span>Hello World!</span>😋L😋l😋L😋) > readableStreamToBlob(stream)
✓ server.browser(<span>Hello World!</span>😋L😋l😋L😋) > readableStreamToArray(stream)
✓ server.browser(<span>Hello World!</span>😋L😋l😋L😋) > readableStreamToArrayBuffer(stream)
✓ server.browser(<span>Hello World!</span>😋L😋l😋L😋) > for await (chunk of stream)
✓ server.browser(<span>Hello World!</span>😋L😋l😋L😋) > for await (chunk of stream) (arrayBuffer)

 343 pass
 0 fail
 345 expect() calls
 Ran 343 tests across 1 files [2.13s]
```

<!--
  Demonstrate the code is solid. Example: The exact commands you ran and their output, screenshots / videos if the pull request changes the user interface.
  How exactly did you verify that your PR solves the issue you wanted to solve?
  If you leave this empty, your PR will very likely be closed.
-->

gnoff commented on scripts/rollup/build.js:
```diff
@@ -363,6 +374,7 @@ function getPlugins(
     isUMDBundle && entry === 'react-art' && commonjs(),
     // Apply dead code elimination and/or minification.
     isProduction &&
+      !skipClosureTranspilation &&
```
what's the reason for skipping closure for bun? beyond just minifying there is inlining and code elimination uses for closure. For instance in the prod bun build the `enableFloat` flag is retained in a few places when they would be elided with the closure step.

colinhacks replied:
I've changed this to use closure for the prod build. I was running into issues with resolving bare module specifiers 
 `import ... from "react"`) when the rollup output format is `es`. This isn't a problem when targeting `cjs` since  `process_common_js_modules` is `false`. Switching to a `cjs` target solved the problem, and Bun can handle both formats anyway...ESM was just a nice-to-have.

gnoff replied:
Ah, interesting. are there any material benefits to the esm option or just that it is "the future" and would be nice to use it b/c Bun supports it?

colinhacks replied:
Mostly that it's "the future" :) But Bun transpiles every file anyway so perf is about the same regardless.

gnoff reviewed (changes requested):
An exciting PR! I think the build options need to be considered some more. Also there are runtime specific tests for node/browser to ensure key apis work in every environment. Can you port those tests to a Bun specific implementation?

gnoff commented on packages/react-dom/server.bun.js:
```diff
@@ -0,0 +1,19 @@
+import ReactVersion from 'shared/ReactVersion';
+export {ReactVersion as version};
+
+export function renderToReadableStream() {
+  return require('./src/server/ReactDOMFizzServerBrowser').renderToReadableStream.apply(
+    this,
+    arguments,
+  );
+}
```
cc @sebmarkbage what are our plans for legacy server apis? I would imagine we should support every runtime equally so while we have support for say `renderToString` in Node/Browser we should also have that in Bun, and then when we drop those we will remove them equally from each runtime

colinhacks replied:
I've added `renderToString` and `renderToStaticMarkup` to `server.bun.js`.

eps1lon commented:
> The built versions of build/node_modules/react-dom/esm/react-dom-server.bun.{development|production}.js have been tested against Bun's React test suite

FYI you can also test against built artifacts from this PR: https://react-builds.vercel.app/api/prs/25597/packages/react-dom

You can also test against a specific commit. https://react-builds.vercel.app/ explains this in more detail

colinhacks commented:
Thanks for the comments @gnoff! I've addressed your reviews and comments. 

Regarding tests: I could port `ReactDOMFizzServerBrowser-test.js` to Bun, but the tests would fail unless executed with the Bun runtime, as they rely on a Bun-specific API in `ReadableStream`:

```ts
new ReadableStream({ type: 'direct' })
```

Since all tests are currently executed with `node`, this would fail for `renderToReadableStream`. I figured I should ask before including Bun in the CI environment, since that may not be appreciated, but I think it's doable. We'll have tests in Bun that will run against `react-dom@next` so we'd catch regressions. Let me know what you think is best 👍

gnoff commented:
@colinhacks Ah, that's right, i forgot that our browser test is kinda cheating b/c it is running in a mock web standards environment with Jest. You're right it doesn't make sense to add the tests unless we can actually run it in the propert runtime. I'm going to look into this b/c we ought to do tests in all the runtimes really. If you're so included and want to hack on it as well to get something working for Bun too that would be cool. I'll follow up in a few days when I know if this is something I'm acutally going to prioritize right now-ish

gnoff commented on packages/react-dom/server.bun.js:
```diff
@@ -11,8 +11,22 @@
 import ReactVersion from 'shared/ReactVersion';
 export {ReactVersion as version};
 
+export function renderToString() {
+  return require('./src/server/ReactDOMLegacyServerBrowser').renderToString.apply(
+    this,
+    arguments,
+  );
+}
+
+export function renderToStaticMarkup() {
+  return require('./src/server/ReactDOMLegacyServerBrowser').renderToStaticMarkup.apply(
+    this,
+    arguments,
+  );
+}
+
```
We should add the other two. you can see how we handle this with .browser.js here

https://github.com/facebook/react/blob/main/packages/react-dom/src/server/ReactDOMLegacyServerBrowser.js#L42-L54

basically we want all exports to be exported on all variants so if you end up using the wrong API in the wrong environment you get a useful error rather than some invalid import error.

gnoff commented on packages/react-reconciler/src/forks/ReactFiberHostConfig.bun.js:
```diff
@@ -0,0 +1,10 @@
+/**
+ * Copyright (c) Meta Platforms, Inc. and affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * @flow
+ */
+
+export * from 'react-dom-bindings/src/client/ReactDOMHostConfig';
```
I'm not sure if you've already explored this and found it was required (I'm not terribly familiar with all the nuances of the forking implementation we have) but I am surprised the HostConfig needs a bun fork since this code doesn't ever run in the Bun runtime.

Presumably it is because the react-dom/client bundle it part of `BUN_DEV` and `BUN_PROD`? I am guessing there is a way to not need

cc @sebmarkbage

colinhacks replied:
I'm not an expert either, but it seems that once you go down the path of adding a new fork, you need to shim everything for Flow to be happy. This is because the generated Flow configs remap a bunch of imports unconditionally: https://github.com/colinhacks/react/blob/add-bun-fizz-server/scripts/flow/createFlowConfigs.js#L56-L61

Interestingly I can `return null` from the `forks.js` resolver for `react-reconciler/src/ReactFiberHostConfig.js` and the generated bundle doesn't change, so you're correct that it isn't being bundled. But I think this is the path of least resistance until the forking gets a little more clever (or doesn't - a few unnecessary re-exports never hurt anyone 😛)

gnoff commented on packages/react-server/src/ReactServerStreamConfigBun.js:
```diff
@@ -0,0 +1,86 @@
+
+// For now we support AsyncLocalStorage as a global for the "browser" builds
+// TODO: Move this to some special WinterCG build.
+// export const supportsRequestStorage = typeof AsyncLocalStorage === 'function';
+// export const requestStorage: AsyncLocalStorage<
+//   Map<Function, mixed>,
+// > = supportsRequestStorage ? new AsyncLocalStorage() : (null: any);
```
is AsyncLocalStorage on the roadmap? I would generally say just delete the comments since they can be referenced from other implementations later if you add this back in

colinhacks replied:
Not in the near term. I'll delete those comments - good call

gnoff reviewed (changes requested):
Thanks! the other changes look good. I left a few more comments and also see the CLA still needs to be completed for this to be mergable

Jarred-Sumner commented:
@gnoff What's the next step for the CLA? I signed it yesterday + added me & @colinhacks under "Schedule A"

Jarred-Sumner commented:
## Performance numbers

For a "hello world" component served over http on an M1 in Bun v0.2.2

| HTTP req/s | ReactDOM renderer |
|------------|----------|
| 102,321 | `react-dom/server.bun` |
| 29,232 | `react-dom/server.browser` |

Overall, this makes SSR'ing React sent over HTTP/HTTPS in Bun up to **3.5x faster**, without changing the public API interface. 

When using `new Response(stream).arrayBuffer()` or `new Response(stream).text()` its closer to 2x (we invested more time optimizing it for HTTP/HTTPS than buffered versions). 

There's nothing React-specific about the optimizations in this PR. The performance win comes from Bun's `"direct"` ReadableStream implementation, which means other frameworks/rendering libraries will get similar performance benefits in Bun for using the `type: "direct"` ReadableStream.

<details>

<summary><code>react-dom/server.bun</code></summary>

```zig
❯ oha http://localhost:3001 -z 5s -c 20
Summary:
  Success rate:	1.0000
  Total:	5.0012 secs
  Slowest:	0.0077 secs
  Fastest:	0.0000 secs
  Average:	0.0002 secs
  Requests/sec:	102321.1759

  Total data:	42.46 MiB
  Size/request:	87 B
  Size/sec:	8.49 MiB

Response time histogram:
  0.000 [10001]  |■
  0.000 [40811]  |■■■■■■■
  0.000 [175510] |■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■
  0.000 [178298] |■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■
  0.000 [77296]  |■■■■■■■■■■■■■
  0.000 [19933]  |■■■
  0.000 [5344]   |
  0.000 [1478]   |
  0.000 [402]    |
  0.001 [113]    |
  0.001 [2542]   |

Latency distribution:
  10% in 0.0001 secs
  25% in 0.0002 secs
  50% in 0.0002 secs
  75% in 0.0002 secs
  90% in 0.0003 secs
  95% in 0.0003 secs
  99% in 0.0004 secs

Details (average, fastest, slowest):
  DNS+dialup:	0.0012 secs, 0.0007 secs, 0.0016 secs
  DNS-lookup:	0.0001 secs, 0.0000 secs, 0.0002 secs

Status code distribution:
  [200] 511728 responses
```

</details>

<details> 

<summary><code>react-dom/server.browser</code></summary>



```zig
❯ oha http://localhost:3001 -z 5s -c 20
Summary:
  Success rate:	1.0000
  Total:	5.0009 secs
  Slowest:	0.0079 secs
  Fastest:	0.0001 secs
  Average:	0.0007 secs
  Requests/sec:	29232.7949

  Total data:	12.13 MiB
  Size/request:	87 B
  Size/sec:	2.43 MiB

Response time histogram:
  0.000 [38]    |
  0.000 [23187] |■■■■■■■■■■■■
  0.001 [40043] |■■■■■■■■■■■■■■■■■■■■■
  0.001 [59202] |■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■
  0.001 [14094] |■■■■■■■
  0.001 [1711]  |
  0.001 [1762]  |
  0.002 [3407]  |■
  0.002 [1941]  |■
  0.002 [395]   |
  0.002 [410]   |

Latency distribution:
  10% in 0.0004 secs
  25% in 0.0005 secs
  50% in 0.0007 secs
  75% in 0.0008 secs
  90% in 0.0009 secs
  95% in 0.0012 secs
  99% in 0.0016 secs

Details (average, fastest, slowest):
  DNS+dialup:	0.0007 secs, 0.0006 secs, 0.0008 secs
  DNS-lookup:	0.0000 secs, 0.0000 secs, 0.0001 secs

Status code distribution:
  [200] 146190 responses
```
</details>


Source code (run with `bun --jsx-production ./react-hello-world.jsx`):

```jsx
import { renderToReadableStream } from "../../test/bun.js/react-dom-server.bun";
const headers = {
  headers: {
    "Content-Type": "text/html",
  },
};

const App = () => (
  <html>
    <body>
      <h1>Hello World</h1>
      <p>This is an example.</p>
    </body>
  </html>
);

const port = Number(process.env.PORT || 3001);
Bun.serve({
  port,
  async fetch(req) {
    return new Response(await renderToReadableStream(<App />), headers);
  },
});
```

gnoff commented on packages/react-dom/package.json:
```diff
@@ -40,12 +41,14 @@
     "cjs/",
-    "umd/"
+    "umd/",
+    "esm/"
```
I think this is technically no longer needed since the bun builds now use the cjs output

colinhacks replied:
Removed!

gnoff reviewed (approved):
I think this is good now. I had one more nit re: package.json

@sebmarkbage I haven't made progress on changing our testing strategy meaningfully for server runtimes so no imminent solution to actually running these tests in CI in a Bun runtime. Is solving the testing issue a blocker to merging this in your mind?

sebmarkbage commented:
We can probably land without tests for now.
